contracts/src/Staking.sol

```diff
@@ -0,0 +1,476 @@
+    event TokensRecovered(address indexed token, address indexed to, uint256 amount);
+
+    // ============================================================
+    // ERRORS
+    // ============================================================
+
+    /*
+     * @notice Thrown when an amount parameter is 0 or invalid
+     */
+    error InvalidAmount();
+
+    /*
+     * @notice Thrown when an address parameter is the zero address
+     */
+    error InvalidAddress();
+
+    /*
+     * @notice Thrown when attempting to stake to a non-registered validator
+     */
+    error NotValidator();
+
+    /*
+     * @notice Thrown when trying to withdraw more than the current stake
+     */
+    error InsufficientStake();
+
+    /*
+     * @notice Thrown when trying to execute a proposal that hasn't been set
+     */
+    error ProposalNotSet();
+
+    /*
+     * @notice Thrown when trying to execute a proposal with an invalid hash
+     */
+    error InvalidProposalHash();
+
+    /*
+     * @notice Thrown when trying to execute a proposal before the timelock expires
+     */
+    error ProposalNotExecutable();
+
+    /*
+     * @notice Thrown when trying to execute a non-existent proposal
+     */
+    error NoProposalExists();
+
+    /*
+     * @notice Thrown when trying to recover more tokens than available
+     */
+    error InsufficientRecoverableAmount();
+
+    /*
+     * @notice Thrown when trying to claim from an empty withdrawal queue
+     */
+    error WithdrawalQueueEmpty();
+
+    /*
+     * @notice Thrown when trying to claim a withdrawal that isn't ready
+     */
+    error NoClaimableWithdrawal();
+
+    /*
+     * @notice Thrown when input arrays have mismatched lengths
+     */
+    error ArrayLengthMismatch();
+
+    /*
+     * @notice Thrown when a parameter is outside acceptable bounds
+     */
+    error InvalidParameter();
+
+    // ============================================================
+    // CONSTRUCTOR
+    // ============================================================
+
+    constructor(address initialOwner, address _safeToken, uint128 initialWithdrawDelay, uint256 _configTimeDelay)
+        Ownable(initialOwner)
+    {
+        if (_safeToken == address(0)) revert InvalidAddress();
+        if (initialWithdrawDelay == 0) revert InvalidParameter();
+        if (_configTimeDelay == 0) revert InvalidParameter();
+
+        safeToken = IERC20(_safeToken);
+        configTimeDelay = _configTimeDelay;
+        withdrawDelay = initialWithdrawDelay;
+        nextWithdrawalId = 1;
+    }
+
+    // ============================================================
+    // EXTERNAL FUNCTIONS - STAKING OPERATIONS
+    // ============================================================
+
+    /*
+     * @notice Stake tokens toward a validator
+     * @param validator The validator address to stake toward
+     * @param amount The amount of tokens to stake
+     */
+    function stake(address validator, uint256 amount) external {
```

Comment:
Interesting, so all stake is delegated? I like it! It makes distinct paths/concepts in the contract.